auth/options.go

```diff
@@ -32,19 +32,28 @@ type Option func(*Options)
 // Options contains options for configuring the behavior of the provider methods.
 // Not all providers/methods support all options.
 type Options struct {
-	Client          client.Client
-	Cache           *cache.TokenCache
-	ServiceAccount  *client.ObjectKey
-	InvolvedObject  cache.InvolvedObject
-	Audiences       []string
-	Scopes          []string
-	STSRegion       string
-	STSEndpoint     string
-	ProxyURL        *url.URL
-	CAData          string
-	ClusterResource string
-	ClusterAddress  string
-	AllowShellOut   bool
+	Client                  client.Client
+	Cache                   *cache.TokenCache
+	ServiceAccountName      string
+	ServiceAccountNamespace string
+	InvolvedObject          cache.InvolvedObject
+	Audiences               []string
+	Scopes                  []string
+	STSRegion               string
+	STSEndpoint             string
+	ProxyURL                *url.URL
+	CAData                  string
+	ClusterResource         string
+	ClusterAddress          string
+	AllowShellOut           bool
+}
+
+// ShouldGetServiceAccountToken returns true if ServiceAccount token should be retrieved.
+func (o *Options) ShouldGetServiceAccountToken() bool {
+	// ServiceAccount namespace is required because ServiceAccounts are namespace-scoped resources.
+	// ServiceAccountName can be empty as it may be provided by DEFAULT_SERVICE_ACCOUNT or
+	// DEFAULT_KUBECONFIG_SERVICE_ACCOUNT environment variables.
+	return o.ServiceAccountNamespace != ""
```

Comment:
This logic is incomplete. If neither `o.ServiceAccountName` nor the env vars are set, we should not fetch a service account token.

```suggestion
	return o.ServiceAccountNamespace != "" &&
		(o.ServiceAccountName != "" || hasDefaultServiceAccount())
```

And:

```go
func hasDefaultServiceAccount() {
	// Here we can detect a default service account by checking either the default
	// service account or the default kubeconfig service account because these two
	// are supposed to never be set simultaneously. The controller main functions
	// must ensure this property.
	return GetDefaultServiceAccount() != "" || GetDefaultKubeconfigServiceAccount() != ""
}
```

Reply:
Edited a little bit.